index.go

```diff
@@ -219,64 +299,52 @@ func (w *indexWriter) writeSeries() error {
 	}
 	sort.Sort(series)
 
-	// Current end of file plus 5 bytes for section header.
-	// TODO(fabxc): switch to relative offsets.
-	base := uint32(w.n) + 5
-
-	w.b = w.b[:0]
-	buf := make([]byte, binary.MaxVarintLen64)
+	// Header holds number of series.
+	w.buf1.reset()
+	w.buf1.putBE32int(len(series))
+	if err := w.write(w.buf1.get()); err != nil {
+		return errors.Wrap(err, "write series count")
+	}
 
 	for _, s := range series {
-		// Write label set symbol references.
-		s.offset = base + uint32(len(w.b))
-
-		n := binary.PutUvarint(buf, uint64(len(s.labels)))
-		w.b = append(w.b, buf[:n]...)
+		w.buf2.reset()
+		w.buf2.putUvarint(len(s.labels))
 
 		for _, l := range s.labels {
-			n = binary.PutUvarint(buf, uint64(w.symbols[l.Name]))
-			w.b = append(w.b, buf[:n]...)
-			n = binary.PutUvarint(buf, uint64(w.symbols[l.Value]))
-			w.b = append(w.b, buf[:n]...)
+			w.buf2.putUvarint32(w.symbols[l.Name])
+			w.buf2.putUvarint32(w.symbols[l.Value])
 		}
 
-		// Write chunks meta data including reference into chunk file.
-		n = binary.PutUvarint(buf, uint64(len(s.chunks)))
-		w.b = append(w.b, buf[:n]...)
+		w.buf2.putUvarint(len(s.chunks))
 
 		for _, c := range s.chunks {
-			n = binary.PutVarint(buf, c.MinTime)
-			w.b = append(w.b, buf[:n]...)
-			n = binary.PutVarint(buf, c.MaxTime)
-			w.b = append(w.b, buf[:n]...)
+			w.buf2.putVarint64(c.MinTime)
+			w.buf2.putVarint64(c.MaxTime)
+			w.buf2.putUvarint64(c.Ref)
 
-			n = binary.PutUvarint(buf, uint64(c.Ref))
-			w.b = append(w.b, buf[:n]...)
+			w.crc32.Reset()
+			c.hash(w.crc32)
+			w.buf2.putBytes(w.crc32.Sum(nil))
 		}
-	}
 
-	return w.section(len(w.b), flagStd, func(wr io.Writer) error {
-		return w.write(wr, w.b)
-	})
-}
+		s.offset = uint32(w.pos)
 
-func (w *indexWriter) init() error {
-	if err := w.writeSymbols(); err != nil {
-		return err
-	}
-	if err := w.writeSeries(); err != nil {
-		return err
+		w.buf1.reset()
+		w.buf1.putUvarint(w.buf2.len())
+
+		w.buf2.putHash(w.crc32)
+
+		if err := w.write(w.buf1.get(), w.buf2.get()); err != nil {
+			return errors.Wrap(err, "write series data")
+		}
 	}
-	w.started = true
 
 	return nil
```

Comment:
So the 4GB limit imposed above has been constantly been on my mind. If the serialised index can be greater than 4GB, the it is highly likely that the series labels and chunks info inside `w.series` would be a significant value (>2GB by a very conservative estimate?).

But we are keeping this info around even when it is not required. Having another map with `[ref]offset` would give us significant savings as we will be able to purge `w.series` after this call. I am not sure if this optimisation is worth it (it has been bugging me since last night :smile: ) but if it is, I think a TODO would be good here.